Alfalfa Job Refactor

TShapinsky opened the pull request:
- Job class added
- Preliminary migration of simulation classes
- Shims in dispatcher to call jobs
- Messages handling with responses and first pass error handling
- Logs generated in a Job are logged to all Runs they own
- Runs are created from a model and maintain a `runs/id.tar.gz` s3 location
- Runs remember which Jobs have been run on them

TODO:

- [x] Make status reporting consistent so mongo_db calls can be moved out of jobs
- [ ] Refactor created step_run jobs to make them cleaner/align fmu and osw run jobs
- [x] Decide how status should be reported and incremented (does each run go from starting to stopped per job? Are job and run statuses different?)
- [ ] Checkin errored runs, possibly to a different s3 location
- [x] Create a mock job runner construct for unit testing jobs

Current Folder sturcture:
`/runs/{run.id}`

kbenne commented on alfalfa_worker/lib/job.py:
```diff
@@ -0,0 +1,226 @@
+    return func
+
+
+class JobMetaclass(type):
```
Maybe some comments to convey what the main purpose is.

TShapinsky replied:
I definitely agree. There will be more comments and doc strings in the next revision

TShapinsky commented on alfalfa_worker/lib/job.py:
```diff
@@ -0,0 +1,226 @@
+
+    def _set_status(self, status: "JobStatus"):
+        self.logger.info(f"Job Status: {status.name}")
+        # A callback could be added here to tell the client what the status is
```
The thing about messages right now is that they aren't received unless the message loop is running. This means that right now if that was added you would only be able to query status in certain jobs after they run the message loop (so not while they're advancing or running anything else)

TShapinsky replied:
This functionality would require some sort of loop to allow responding while execution is happening.

TShapinsky commented on alfalfa_worker/lib/job.py:
```diff
@@ -0,0 +1,226 @@
+        """Runs job
+        called by start()"""
+
+    @message
```
Yes the return of the message is jsonified and sent back over redis

TShapinsky commented on alfalfa_worker/dispatcher.py:
```diff
@@ -135,3 +161,38 @@ def run(self):
+
+    def start_job(self, job_name, parameters):
+        """Start job in thread by Python class path"""
+        klazz = self.find_class(job_name)
```
Yes, that is the goal I think. Right now I didn't want to modify the web server to send the corresponding messages with job names.

anyaelena replied:
is the plan to 
1. merge in updates to the worker/jobs w/out changes to web, and then
2. coordinate web updates and worker cleanup?

This seems reasonable to me provided there isn't too much overhead to avoid changes in the worker.  

@kbenne at what point would you be able to start integrating w/BOPtest?  1 or 2 above?

TShapinsky commented on alfalfa_worker/dispatcher.py:
```diff
@@ -99,7 +109,15 @@ def process_message(self, message):
                         # worker class that was assigned
                         return worker_class
                     elif action == 'addSite':
```
The conditional switches will be removed. But changes need to be made on the web end to send the right messages.

TShapinsky replied:
@kbenne if you check now you will see that the dispatcher is now very slim

nllong commented on alfalfa_worker/jobs/openstudio/create_run.py:
```diff
@@ -0,0 +1,118 @@
+            submitted_osw_path = osws[0]
+            submitted_workflow_path = os.path.dirname(submitted_osw_path)
+        else:
+            raise JobExceptionInvalidModel("No .osw file found")
```
nice... i like the named exception!

nllong commented on alfalfa_worker/jobs/openstudio/create_run.py:
```diff
@@ -0,0 +1,118 @@
+            raise JobExceptionInvalidModel("No .osw file found")
+
+        # locate the "default" workflow
+        default_workflow_path: str = os.path.join(lib_dir, 'workflow/workflow.osw')
```
```suggestion
        default_workflow_path: str = os.path.join(lib_dir, 'workflow', 'workflow.osw')
```

nllong commented on alfalfa_worker/jobs/openstudio/create_run.py:
```diff
@@ -0,0 +1,118 @@
+        # run workflow
+        call(['openstudio', 'run', '-m', '-w', submitted_osw_path])
+
+        points_json_path = os.path.join(submitted_workflow_path, 'reports/haystack_report_haystack.json')
```
```suggestion
        points_json_path = os.path.join(submitted_workflow_path, 'reports', 'haystack_report_haystack.json')
```

nllong commented on alfalfa_worker/jobs/openstudio/create_run.py:
```diff
@@ -0,0 +1,118 @@
+        call(['openstudio', 'run', '-m', '-w', submitted_osw_path])
+
+        points_json_path = os.path.join(submitted_workflow_path, 'reports/haystack_report_haystack.json')
+        mapping_json_path = os.path.join(submitted_workflow_path, 'reports/haystack_report_mapping.json')
```
```suggestion
        mapping_json_path = os.path.join(submitted_workflow_path, 'reports', 'haystack_report_mapping.json')
```

Would be nice to use `pathlib.Path`...

nllong commented on alfalfa_worker/jobs/openstudio/create_run.py:
```diff
@@ -0,0 +1,118 @@
+        self.insert_os_tags(points_json_path, mapping_json_path)
+
+        # create a "simulation" directory that has everything required for simulation
+        simulation_dir = self.join('simulation/')
```
remove appended slash and handle elsewhere.

nllong commented on alfalfa_worker/jobs/openstudio/create_run.py:
```diff
@@ -0,0 +1,118 @@
+        simulation_dir = self.join('simulation/')
+        os.mkdir(simulation_dir)
+
+        idf_src_path = os.path.join(submitted_workflow_path, 'run/in.idf')
```
break out this into its own method.

`sym_link_files(os.path.join(submitted_workflow_path, 'run/in.idf'), os.path.join(simulation_dir, 'sim.idf'))`

nllong commented on alfalfa_worker/jobs/openstudio/full_run.py:
```diff
@@ -0,0 +1,24 @@
+        for oswpath in osws:
+            subprocess.call(['openstudio', 'run', "-w", oswpath])
+
+        time = str(datetime.now(tz=pytz.UTC))
```
eek, timezones

TShapinsky replied:
yeah... I just copied the pattern we were using elsewhere. If we want to address that it is definitely a possibility

nllong commented on alfalfa_worker/jobs/openstudio/full_run.py:
```diff
@@ -0,0 +1,24 @@
+            subprocess.call(['openstudio', 'run', "-w", oswpath])
+
+        time = str(datetime.now(tz=pytz.UTC))
+        self.mongo_db_sims.update_one({"_id": self.run.id}, {"$set": {"simStatus": "Complete", "timeCompleted": time, "s3Key": f"run/{self.run.id}.tar.gz"}}, False)
```
It will be nice to either 1) remove mongo or 2) use a ODM to talk to mongo. Regardless, no actions are needed for this PR.

nllong commented on alfalfa_worker/lib/auto_name_enum.py:
```diff
@@ -0,0 +1,9 @@
+from enum import Enum
+
+
+class AutoName(Enum):
```
can you document this?

TShapinsky replied:
yeah. I'll add a docstring. The gist is that it changed the `auto()` method so that the generated values are just the name.

nllong commented on alfalfa_worker/lib/data/data_generator.py:
```diff
@@ -172,7 +172,7 @@ def generate_weather(self,
 
         # Change to Resources directory
-        currdir = os.curdir
+        currdir = os.getcwd()
```
would love to get rid of these types of calls!

nllong commented on alfalfa_worker/lib/job.py:
```diff
@@ -0,0 +1,265 @@
+                __old_init__(self, *args, **kwargs)
+            self.set_job_status(JobStatus.INITIALIZED)
+        cls_dicts['__init__'] = __new_init__
+        klazz = super().__new__(cls, name, bases, cls_dicts)
```
see comment above about pep8 converntions

nllong commented on alfalfa_worker/lib/job.py:
```diff
@@ -0,0 +1,265 @@
+        return klazz
+
+
+class Job(metaclass=JobMetaclass):
```
we should break these out into their own classes if possible.

TShapinsky replied:
I'm not sure what you mean by this

nllong commented on alfalfa_worker/lib/point.py:
```diff
@@ -0,0 +1,46 @@
+
+
+class Point:
+    """Input or Ouput point used to exchange data between simulation and client"""
```
```suggestion
    """Input or Output point used to exchange data between simulation and client"""
```

nllong commented on alfalfa_worker/lib/point.py:
```diff
@@ -0,0 +1,46 @@
+from uuid import uuid4
+
+
+class Point:
```
I like the use of `Point`!

nllong commented on alfalfa_worker/lib/point.py:
```diff
@@ -0,0 +1,46 @@
+
+
+class PointType(Enum):
+    INPUT = auto()
```
I'm not really familiar with the `auto()` type. Looks helpful!

nllong commented on alfalfa_worker/lib/run.py:
```diff
@@ -0,0 +1,91 @@
+        self.sim_time = sim_time
+        self.points: List[Point] = []
+
+    def join(self, *args):
```
we could move this to pathlib.Path...

TShapinsky replied:
as in, the run would expose a path object? so people would do `run.dir / 'my.osw'`?

nllong replied:
yeah, that is what I'm thinking...

nllong commented on alfalfa_worker/lib/run_manager.py:
```diff
@@ -0,0 +1,182 @@
+from alfalfa_worker.lib.sim_type import SimType
+
+
+class RunManager(LoggerMixinBase):
```
love it

nllong reviewed (commented):
this is a great! You have done a large amount of work to abstract out (via metaclasses, mixins, and inheritance) a structure that will support multiple job types across multiple engines! 

Few high level items:

* run poetry update
* tests don't pass, but that is probably a known
* docker-compose build failed for me. Not sure what is going on. I have sum mismatches. Not on VPN either.
* use of mypy syntax throughout will be very nice!


My biggest concern is that a lot of great development has happened and has been completed without testing. I think it is really important to build the tests alongside the development of new functionality, otherwise, the intent and staying power of the new functionality is lost.

TShapinsky commented on alfalfa_worker/jobs/modelica/__init__.py:
```diff
@@ -0,0 +1,3 @@
+from .create_run import CreateRun
+from .full_run import FullRun
+from .step_run import StepRun
```
We can definitely move them out. The only reason they're here is to make the job path shorter. So instead of `alfalfa_worker.jobs.openstudio.create_run.CreateRun` you could just run the job `alfalfa_worker.jobs.openstudio.CreateRun`. Not a big deal either way though

TShapinsky commented on alfalfa_web/server/schema.js:
```diff
@@ -149,6 +149,33 @@ const simType = new GraphQLObjectType({
   })
 });
 
+const runType = new GraphQLObjectType({
```
Yes we are. But right now it's the only way to get information to the client. I also figured this could have give Alex some reference implementation for when he is moving it to a REST api.

TShapinsky commented on alfalfa_worker/dispatcher.py:
```diff
@@ -52,67 +53,30 @@ def __init__(self):
-
-        return worker_class
+        self.workdir = '/runs'
+        if not os.path.exists(self.workdir):
```
That sounds good.

TShapinsky commented on alfalfa_worker/jobs/modelica/create_run.py:
```diff
@@ -0,0 +1,127 @@
+                        break
+        return site_ref
+
+    def insert_fmu_tags(self):
```
Yeah, I need to dig into that. One of the next things on my list is to actually integrate `Point` class, which should touch a good amount of the tagging stuff.

TShapinsky commented on alfalfa_worker/jobs/modelica/full_run.py:
```diff
@@ -0,0 +1,34 @@
+from alfalfa_worker.lib.testcase import TestCase
+
+
+class FullRun(AlfalfaConnectionsBase, Job):
```
I think we may have decided to just remove this class since you can do similar with the step run in timescale, though maybe this is not the case. 

If we keep it I'm all for AnnualRun, makes a lot more sense.

TShapinsky commented on alfalfa_web/server/resolvers.js:
```diff
@@ -86,52 +91,32 @@ function runSiteResolver(args) {
-      });
+      }`,
+      QueueUrl: process.env.JOB_QUEUE_URL,
+      MessageGroupId: "Alfalfa"
```
possibly. I just copied this from another function...

TShapinsky commented:
> this is a great! You have done a large amount of work to abstract out (via metaclasses, mixins, and inheritance) a structure that will support multiple job types across multiple engines!
> 
> Few high level items:
> 
>     * run poetry update
> 
>     * tests don't pass, but that is probably a known
> 
>     * docker-compose build failed for me. Not sure what is going on. I have sum mismatches. Not on VPN either.
> 
>     * use of mypy syntax throughout will be very nice!
> 
> 
> My biggest concern is that a lot of great development has happened and has been completed without testing. I think it is really important to build the tests alongside the development of new functionality, otherwise, the intent and staying power of the new functionality is lost.

I definitely agree. I need to add more unit tests. I have been making sure that the integration tests continue to work as I've been doing this development.

TShapinsky commented:
@nllong Do you think you could give the testing infrastructure I've created a look and let me know if you think this is adequate?

anyaelena commented:
@nllong @kbenne  this is ready for review

anyaelena commented:
@TShapinsky Please resolve conflicts now that @axelstudios's PR is merged.

nllong commented on alfalfa_web/components/Sites/Sites.js:
```diff
@@ -104,24 +104,30 @@ class Sites extends React.Component {
   isRemoveButtonDisabled = () => {
     const stoppedItem = this.selectedSites().some((item) => {
-      return item.simStatus === "Stopped";
+      return item.simStatus === "READY" || item.simStatus == "COMPLETE" || item.simStatus == "ERROR";
```
nice

nllong commented on setup.cfg:
```diff
@@ -72,10 +72,11 @@ extras = True
     --cov alfalfa_worker --cov-report term-missing
     --verbose
-    -m "not integration and not fmu"
+    -m "not integration and not fmu and not docker"
```
yikes, lots of markers. Will be nice to reduce this if possible.

nllong reviewed (approved):
This is great work. I am impressed with all the testing through the Mock*! 

I create a PR with some tweaks around dependencies and README text [here](https://github.com/NREL/alfalfa/pull/306).

TShapinsky commented:
> I doubt the "run simulation tests" is going to finish due to timeouts, so we will need to figure out how to reduce the runtimes of those models.
> 
> I added Release Instructions to our Wiki https://github.com/NREL/alfalfa/wiki/Release-Instructions. They probably need some tweaking, but it is a start.
> 
> I've also updated the alfalfa-client a bit more and prepared a PR for release (this will be version 0.3.0) [NREL/alfalfa-client#33](https://github.com/NREL/alfalfa-client/pull/33)
> 
> I've added a brief description to Docker Hub's images--feel free to update.
> 
> I think these are some of the remaining steps
> 
>     * [ ]  Verify version of next release and update pypoetry.toml
> 
>     * [ ]  Update CHANGELOG.md
> 
>     * [ ]  Release alfalfa-client
> 
>     * [ ]  Unpin alfalfa-cient branch in pypoetry.tml and update with latest of pypi's alfalfa-client

I think what is happening with the run simulation tests is: because it has to spin up a thread for the job to run in, if the test fails it doesn't exit the thread and thus kinda spins forever... So it's not a model runtime issue but more of a pytest

TShapinsky commented:
The weird thing about the failures in these tests is that the `Run simulation tests (push)` is consistently failing while the `Run simulation tests (pull_request)` passes...

TShapinsky commented:
@nllong 
This is the log from the error that is happening (I created a run which just calls advance until it skips a step and logs kStep values):
```
INFO     StepRun:step_run.py:348 setting up points
INFO     StepRun:step_run.py:105 current_ep_time: 2019-01-02 00:00:00, start_datetime: 2019-01-02 00:02:00
INFO     StepRun:step_run.py:110 current_ep_time: 2019-01-02 00:00:00, start_datetime: 2019-01-02 00:02:00
INFO     StepRun:step_run.py:138 Incrementing kStep from 1
INFO     StepRun:step_run.py:141 kStep is now: 2
INFO     StepRun:step_run.py:110 current_ep_time: 2019-01-02 00:01:00, start_datetime: 2019-01-02 00:02:00
INFO     StepRun:step_run.py:138 Incrementing kStep from 2
INFO     StepRun:step_run.py:141 kStep is now: 3
INFO     StepRun:step_run.py:113 current_ep_time: 2019-01-02 00:02:00 reached desired start_datetime: 2019-01-02 00:02:00
INFO     StepRun:step_run.py:301 updating db time to: s:2019-01-02 00:02:00
INFO     StepRun:step_run_base.py:88 Running external_clock
INFO     StepRun:job.py:171 Job Status: WAITING
INFO     StepRun:job.py:192 recieved message: {'type': 'message', 'pattern': None, 'channel': b'f70c8631-2011-4bff-aa73-9e89a2cd108b', 'data': b'{"method": "advance", "message_id": "44ed6fda-d43b-4ec9-a109-972638e746ea", "params": {}}'}
INFO     StepRun:job.py:171 Job Status: RUNNING
INFO     StepRun:step_run.py:372 advance called at time 2019-01-02 00:02:00
INFO     StepRun:step_run.py:138 Incrementing kStep from 3
INFO     StepRun:step_run.py:141 kStep is now: 4
INFO     StepRun:step_run.py:301 updating db time to: s:2019-01-02 00:03:00
INFO     StepRun:job.py:217 message_id: 44ed6fda-d43b-4ec9-a109-972638e746ea, response: {'response': None, 'status': 'ok'}
INFO     StepRun:job.py:171 Job Status: WAITING
INFO     StepRun:job.py:192 recieved message: {'type': 'message', 'pattern': None, 'channel': b'f70c8631-2011-4bff-aa73-9e89a2cd108b', 'data': b'{"method": "advance", "message_id": "54f85325-54a9-4a81-bdd0-1a2d0aae7b3c", "params": {}}'}
INFO     StepRun:job.py:171 Job Status: RUNNING
INFO     StepRun:step_run.py:372 advance called at time 2019-01-02 00:03:00
INFO     StepRun:step_run.py:138 Incrementing kStep from 4
INFO     StepRun:step_run.py:141 kStep is now: 5
INFO     StepRun:step_run.py:301 updating db time to: s:2019-01-02 00:05:00
INFO     StepRun:job.py:217 message_id: 54f85325-54a9-4a81-bdd0-1a2d0aae7b3c, response: {'response': None, 'status': 'ok'}
INFO     StepRun:job.py:171 Job Status: WAITING
```
It looks like sometimes the simulation is advancing itself two steps instead of one, I have no idea why this is happening and seems to be happening randomly... Any ideas?

TShapinsky commented:
Notes before I leave:
1. I'm not sure if the flakiness in the simulation test is a result of this PR it seems like there could be something going on with mlep, bcvtb or even the test building. I have gotten this error when running the `simple_external` integration test before though it seems to happen more often with the simulation test.
 i. The best success I've had reproducing this bug locally is to increase the number of times the test increments the timestep (as the error seems to happen randomly). That can be done by modifying [this line](https://github.com/NREL/alfalfa/blob/job_refactor/tests/jobs/openstudio/test_openstudio_workflow.py#L74)
 ii. To run the specific problem test start the service with `docker-compose -f docker-compose.yml -f docker-compose.dev.yml up --build -d` and then run the test with `docker exec alfalfa_worker_1 bash -c "cd /alfalfa && pytest -m docker tests/jobs/openstudio -k test_simple_external_clock -o log_cli=true -o log_level=info"`
2. One thing which I was trying to make better is the fact that certain tests which use threads to simulate a job running concurrently to the client will spin forever if they fail (the job is never stopped). This will only happen if the test fails and will cause github actions to eventually cancel the job anyways so it's mostly a nuisance. A hotfix to this would be applying a timeout in the github action until we can figure out how to guarantee thread or process termination on a failure.

anyaelena commented:
Because the simulation tests are new coverage with this PR and we haven't determined whether the nondeterministic clock advance failures are real or an artifact of the testing setup, I've [commented](https://github.com/NREL/alfalfa/pull/299/commits/8639bd387c7496a8f9b3610ff3f44a9dcab65ab1) the assertions that were (sometimes) failing and opened #308 to revisit.  For now, I'm merging this PR so we can move forward with manual testing/qa for the 0.3.0 release.

TShapinsky commented:
I think the run simulations test is timing out because when the test fails
there is a thread that just spins forever. When the test succeeds the
thread completes naturally. You can see when it times out that there is a
full pytest log which was written, but the process never exited due to the
unterminated thread.

On Fri, Jun 17, 2022, 08:53 Nicholas Long ***@***.***> wrote:

> I doubt the "run simulation tests" is going to finish due to timeouts, so
> we will need to figure out how to reduce the runtimes of those models.
>
> I added Release Instructions to our Wiki
> https://github.com/NREL/alfalfa/wiki/Release-Instructions. They probably
> need some tweaking, but it is a start.
>
> I've also updated the alfalfa-client a bit more and prepared a PR for
> release (this will be version 0.3.0) NREL/alfalfa-client#33
> <https://github.com/NREL/alfalfa-client/pull/33>
>
> I've added a brief description to Docker Hub's images--feel free to update.
>
> I think these are some of the remaining steps
>
>    - Verify version of next release and update pypoetry.toml
>    - Update CHANGELOG.md
>    - Release alfalfa-client
>    - Unpin alfalfa-cient branch in pypoetry.tml and update with latest of
>    pypi's alfalfa-client
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NREL/alfalfa/pull/299#issuecomment-1158951263>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABDNI6NWBJYKT23HLPHTHEDVPSGPLANCNFSM5VZCB5SQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
